src/scoring/hpf_scoring.py

```diff
@@ -0,0 +1,168 @@
+            self.scoring_region, HPF_output_manifest_id_dict)
+        self.manifest_id_dict = self.datastore.read_json_file(
+            manifest_id_dict_filename)
+        self.manifest_id_dict = {n: set(x)
+                                 for n, x in self.manifest_id_dict.items()}
+        rating_matrix_filename = os.path.join(
+            self.scoring_region, HPF_output_rating_matrix)
+        self.datastore.download_file(
+            rating_matrix_filename, "/tmp/rating_matrix.npy")
+        self.rating_matrix = np.load("/tmp/rating_matrix.npy")
+        beta_matrix_filename = os.path.join(
+            self.scoring_region, HPF_output_item_matrix)
+        self.datastore.download_file(
+            beta_matrix_filename, "/tmp/beta_matrix.npy")
+        self.beta = np.load("/tmp/beta_matrix.npy")
+        self.manifests, self.packages = self.rating_matrix.shape
+        self.rating_mean = self.rating_matrix.mean()
+
+    def predict(self, input_stack):
+        """Prediction function.
+
+        :param input_stack: The user's package list
+        for which companion recommendation are to be generated.
+        :return companion_recommendation: The list of recommended companion packages
+        along with condifence score.
+        :return package_topic_dict: The topics associated with the packages
+        in the input_stack+recommendation.
+        :return missing_packages: The list of packages unknown to the HPF model.
+        """
+        input_id_set = set()
+        missing_packages = []
+        for package_name in input_stack:
+            package_id = self.package_id_dict.get(package_name)
+            if package_id is None:
+                missing_packages.append(package_name)
+        else:
+            input_id_set.add(package_id)
+        # TODO: Check for known-unknown ratio before recommending
+        companion_recommendation, package_topic_dict = self.folding_in(
+            input_id_set)
+        return companion_recommendation, package_topic_dict, missing_packages
+
+    def match_manifest(self, input_id_set):
+        """Find a manifest list that matches user's input package list and return its index.
+
+        :param input_id_set: A set containing package ids of user's input package list.
+        :return manifest_id: The index of the matched manifest.
+        """
+        for manifest_id, dependency_set in self.manifest_id_dict.items():
+            if dependency_set == input_id_set:
+                break
+        else:
+            manifest_id = -1
+        return manifest_id
+
+    def folding_in(self, input_id_set):
+        """Folding in logic for prediction.
+
+        :param  input_id_set: A set containing package ids of user's input package list.
+        :return: Filter companion recommendations and their topics.
+        """
+        manifest_id = int(self.match_manifest(input_id_set))
+        if manifest_id == -1:
+            theta = []
+            with tf.Session() as sess:
```

Comment:
@rootAvish I had a doubt, should the `tf.session()` be created at each request, or at the class level.